### DISCUSSION TITLE
GO plugins are all messed up

### DISCUSSION THREAD

**Discussion by ssozonoff** [lf-edge/ekuiper#3113] [Q&A]
Hi.

So I quickly tried to get a GO plugin up and running using information I found on the internet and the Docker dev image for ekuiper but every time it tries to load my plugin I get an error similar to the following.

`2024-08-19 12:52:39 time="2024-08-19T10:52:39Z" level=error msg="plugin tcp open error: plugin.Open(\"/go/kuiper/_build/kuiper-1.14.1-linux-arm64/plugins/sources/tcp@v1.0.0\"): plugin was built with a different version of package github.com/lf-edge/ekuiper/pkg/api" file="native/manager.go:746"`

Reading some more and you find tons people having similar issues with GO plugins and the consensus seems to be its a GO problem.

I have tried compiling like this

`GO111MODULE=on CGO_ENABLED=1 GOOS=linux GOARCH=arm64  go build -trimpath -ldflags="-s -w" --buildmode=plugin -o /home/target/plugins/sources/tcp@v1.0.0.so sources/tcp.go`

but does not resolve the issue ... so I may try and Python version and see if I can get that working.

Any tips appreciated.

**Comment by ngjaying** [comment]
That is the limitation of go plugin system. There is a tutorial for v1.14 version https://ekuiper.org/docs/en/v1.14/extension/native/develop/plugins_tutorial.html.

**Reply by ssozonoff** [reply]
Thanks I have the plugin installed correctly now using https://ekuiper.org/docs/en/latest/extension/native/develop/plugins_tutorial.html

**Comment by ssozonoff** [comment]
This is just so painful .... so I built the plugin on the same host machine using docker image lfedge/ekuiper:v1.14.1-dev. Then copied the .so file to the second container running lfedge/ekuiper:1.14.1 and the plugin does not load ....

`time="2024-08-20T14:07:11Z" level=error msg="plugin tcp open error: plugin.Open(\"/kuiper/plugins/sources/tcp@v1.0.0\"): plugin was built with a different version of package github.com/lf-edge/ekuiper/pkg/api" file="native/manager.go:746"`

Goes back to what I said in the beginning, the GO plugin stuff is more or less un-usable.

**Reply by ngjaying** [reply]
Plugin built in lfedge/ekuiper:v1.14.1-dev can only be used in lfedge/ekuiper:v1.14.1-slim because the base os is different. P.S. recommend to write the TCP source as a built-in source to save your time
